### PR TITLE
Implement topic and resource creation endpoint in M2 specs

### PR DESCRIPTION
**Pull request by furkanakkurt1335**
I updated the createDB.js according to the new schemas.

I implemented the API specifications for topic and resource creation, given in #352.

### CODE REVIEW

**Review by ersoykadir — CHANGES_REQUESTED** [review]
Reviewed. Actually I just read every line in whole repo. I have an upper understanding of our project now :D 
We are good to go after necessary changes done.

File: app/server/src/models/comment/comment.model.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+const createComment = async (user, body, files) => {
+  var comment = new Comment({
+    user,
+    body,
+    files,
+  });
+
+  const commentTemp = await comment.save();
+  commentTemp.comment_date = commentTemp.createdAt;
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/comment/comment.model.js:21]: I agree!

File: app/server/src/models/discussion/discussion.model.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    user,
+    space,
+    comments,
+    body,
+    files,
+  });
+
+  // Just to set the creation time...
+  const discussionTemp = await discussion.save();
+  discussionTemp.date = discussionTemp.createdAt;
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/discussion/discussion.model.js:41]: No, I agree. We don't need the field if there's an already built-in field for the same thing.

File: app/server/src/models/comment/comment.model.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/comment/comment.model.js]: Done!

File: app/server/src/models/discussion/discussion.model.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+});
+
+discussionSchema.set("timestamps", true);
+const Discussion = mongoose.model("Discussion", discussionSchema);
+
+const createDiscussion = async (user, space, comments, body, files) => {
+  var discussion = new Discussion({
+    user,
+    space,
+    comments,
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/discussion/discussion.model.js]: Done!

File: app/server/src/models/discussion/discussion.model.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/discussion/discussion.model.js]: Done!

File: app/server/src/models/resource/resource.model.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  topic: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: "Topic",
+  },
+  discussion: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: "Discussion",
+  },
+});
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/resource/resource.model.js]: Done!

File: app/server/src/models/topic/topic.model.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    ref: "Badge",
+  },
+  resources: [
+    {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: "Resource",
+    },
+  ],
+});
+
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/topic/topic.model.js:27]: Done!

File: app/server/src/models/enrollment/enrollment.model.js [hunk 6]
```diff
@@ -2,36 +2,37 @@ const mongoose = require("mongoose");
...
-    },
+    }],
     progress: {
       type: Map,
       of: Boolean,
     },
   },
   { timestamps: true }
 );
+enrollmentSchema.index({ user: 1, space: 1 }, { unique: true });
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/enrollment/enrollment.model.js:32]: Sure thing!

File: app/server/src/models/space/space.model.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: "Enrollment",
+    },
+  ],
+  image: {
+    type: String,
+  },
+});
+
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/space/space.model.js:76]: Done

File: app/server/src/routes/badge/badge.controller.js [hunk 8]
```diff
@@ -1,28 +1,23 @@
...
-        }
-        catch (e) {
-            res.status(400).send({ "error": e })
-        }
-    },
+  createBadge: async function (req, res) {
+    try {
+      const { title, description } = req.body;
+      const badge = await BadgeModel.createBadge(title, description);
+      res.status(201).json({ message: badge });
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/badge/badge.controller.js:8]: I don't know, we should just return a message of "Created" right? Or no message, only status code.

File: app/server/src/routes/badge/badge.controller.js [hunk 9]
```diff
@@ -1,28 +1,23 @@
...
-            const badge = await BadgeModel.getBadge(req.params.id)
-            res.status(200).json({ "status": "ok", "message": badge })
-        }
-        catch (e) {
-            res.status(400).send({ "error": e })
-        }
-    },
+  createBadge: async function (req, res) {
+    try {
+      const { title, description } = req.body;
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/badge/badge.controller.js:6]: Badges don't have topics currently, we would need to add it to the model also.

File: app/server/src/models/discussion/discussion.model.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    type: mongoose.Schema.Types.ObjectId,
+    ref: "Space",
+  },
+  comments: [
+    {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: "Comment",
+    },
+  ],
+  body: String,
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/models/discussion/discussion.model.js]: Added!

File: app/server/src/routes/resource/resource.validate.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+const { body, query } = require("express-validator");
+
+exports.validate = (method) => {
+  switch (method) {
+    case "create-resource":
+      return [
+        body("name", "name doesn't exist").exists(),
+        body("body", "body doesn't exist").exists(),
+        body("discussion", "discussion doesn't exist").exists(),
+        body("media", "media doesn't exist").exists().isArray(),
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/resource/resource.validate.js]: Done!

File: app/server/src/routes/discussion/discussion.controller.js [hunk 12]
```diff
@@ -1,32 +1,32 @@
...
+        files
+      );
+      res.status(201).send({ message: discussion });
+    } catch (e) {
+      res.status(400).send({ error: e });
+    }
+  },
+  getDiscussion: async function (req, res) {
+    try {
+      const resource = await DiscussionModel.getPopulatedDiscussion(
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/discussion/discussion.controller.js]: Done!

File: app/server/src/routes/discussion/discussion.controller.js [hunk 13]
```diff
@@ -1,32 +1,32 @@
...
+        req.params.id
+      );
+      res.status(200).json({ message: resource });
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/discussion/discussion.controller.js]: Done!

File: app/server/src/routes/topic/topic.validate.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+const { body, query } = require("express-validator");
+
+exports.validate = (method) => {
+  switch (method) {
+    case "create-topic":
+      return [
+        body("name", "name doesn't exist").exists(),
+        body("badge", "badge doesn't exist").exists(),
+        body("resources", "resources don't exist").exists().isArray(),
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/topic/topic.validate.js]: Done!

File: app/server/src/routes/user/user.validate.js [hunk 15]
```diff
@@ -1,32 +1,71 @@
...
+    case "refresh_tokens": {
+      return [
+        body("email", "Invalid email").exists().isEmail(),
+        body("refresh_token", "Invalid email").exists().isJWT(),
+      ];
+    }
+    case "confirm-email": {
+      return [body("code", "code does not exist").exists().isJWT()];
+    }
+    case "login": {
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/user/user.validate.js]: Done!

File: app/server/src/routes/user/user.validate.js [hunk 16]
```diff
@@ -1,32 +1,71 @@
...
+          .isLength({ min: 8 })
+          .matches(
+            /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})/
+          ),
+      ];
+    }
+    case "refresh_tokens": {
+      return [
+        body("email", "Invalid email").exists().isEmail(),
+        body("refresh_token", "Invalid email").exists().isJWT(),
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/user/user.validate.js]: Done!

File: app/server/src/routes/userProfile/userProfile.controller.js [hunk 17]
```diff
@@ -3,41 +3,50 @@ const personalInfoModel = require("../../models/personalInfo/personalInfo.model"
...
+        knowledge: knowledge,
+      };
+      personalInfoModel.updateBio(personalInfoId, data);
+      res.status(201).send({ message: "Updated bio successfully" });
+    } catch (e) {
+      res.status(400).send({ error: e.toString() });
+    }
+  },
+
+  updatePersonalInfo: async function (req, res) {
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/userProfile/userProfile.controller.js]: Done!

File: app/server/src/routes/user/user.validate.js [hunk 18]
```diff
@@ -1,32 +1,71 @@
...
+      ];
+    }
+    case "confirm-email": {
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/user/user.validate.js:41]: Done!

File: app/server/src/routes/user/user.validate.js [hunk 19]
```diff
@@ -1,32 +1,71 @@
...
+          "Password must contain at least 8 characters, one uppercase, one lowercase, one number and one special case character"
+        )
+          .exists()
+          .isLength({ min: 8 })
+          .matches(
+            /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})/
+          ),
+      ];
+    }
+    case "refresh_tokens": {
```
[the remark this answers is not in the document]
Comment (furkanakkurt1335) [reply app/server/src/routes/user/user.validate.js]: Done!

**Comment by furkanakkurt1335** [comment]
I've gone over all of @ersoykadir 's review. 🙏 I've done commits accordingly.

File: app/server/src/routes/user/user.controller.js [hunk 20]
```diff
@@ -78,9 +78,38 @@ const UserController = {
...
+            } else {
+                return res.status(400).json({
+                    message: "Could not login with the parameters you provided.",
+                });
+            }
+        } catch (error) {
+            return res.status(400).json({
+                message: "Failed to login!",
+                error: error.toString()
+            });
```
Comment (ersoykadir) [inline app/server/src/routes/user/user.controller.js:112]: I don't understand why did you change all these? This should be taken back

**Review by ersoykadir — APPROVED** [review]
It looks good for now.
